### PR TITLE
Add progress bar to lmf.load (#46)

### PR DESCRIPTION
**Pull request by dpalmasan**
Add a progress bar to `lmf.load` (#46 ) function. Since we use generators, as an heuristic we just count lexicon starting tags from the XML source (to reduce memory footprint and to avoid processing twice the generators). For the progress bar I had to add a dependency to `tqdm`.

Here is how it looks like:

Alt Text

### CODE REVIEW

**Review by goodmami — CHANGES_REQUESTED** [review]
Hi, I really appreciate this work, but unfortunately I won't accept it as-is. A few things:

* The `mypy.ini` file is not related to this pull request. If there is an issue running Mypy on different platforms, please open a separate issue.
* Wn provides its own progress bar implementation in [wn.util](https://wn.readthedocs.io/en/latest/api/wn.util.html), so please use that. There at least two reasons:
  1. It avoids a dependency for non-critical functionality.
  2. The code is designed to allow alternative progress bars wrapped in a [wn.util.ProgressHandler](https://wn.readthedocs.io/en/latest/api/wn.util.html#wn.util.ProgressHandler) (or subclass) instance. For instance, [Wordbook](https://github.com/fushinari/wordbook/) hooks in a GUI progress bar. Users can, at their option, use tqdm or something else.
* The method for estimating the file size and updating the progress bar is not reliable (due to the newline issue explained below) and not granular enough.
 
A better solution, perhaps, is to pass along a progress handler (look into `wn.download()` or `wn.add()` for an example of how to instantiate one) to the `wn.lmf.XMLEventIterator` class and update in the `__next__()` method. The counts should come from [lmf.scan_lexicons()](https://wn.readthedocs.io/en/latest/api/wn.lmf.html#wn.lmf.scan_lexicons) (see `wn._add._precheck()` and `wn._add._sum_counts()` to see how those are used).

If you can provide that, I can do the final bit to make sure it's compatible with the old API, etc., which is hard to explain here.

Cheers,

**Comment by dpalmasan** [comment]
Okay, I modified the code per the review, now the result looks like:

Alt Text

Updates include:

* Remove `tdqm` dependency and use `ProgressHandler` utility from the package
* Load on XML read, not only on the lexicons
* Remove unrelated files.

File: wn/lmf.py [hunk 1]
```diff
@@ -565,16 +570,35 @@ def load(source: AnyPath) -> LexicalResource:
...
 
     lexicons: List[Lexicon] = []
-    while events.starts('Lexicon', 'LexiconExtension'):
-        lexicons.append(_load_lexicon(events, version))
-        root.clear()
+    with tqdm(total=_guess_iterations(source)) as pbar:
+        while events.starts('Lexicon', 'LexiconExtension'):
+            lexicons.append(_load_lexicon(events, version))
+            pbar.update()
+            root.clear()
```
[the remark this answers is not in the document]
Comment (dpalmasan) [reply wn/lmf.py]: Addressed

File: wn/lmf.py [hunk 2]
```diff
@@ -565,16 +570,35 @@ def load(source: AnyPath) -> LexicalResource:
...
+    an object with len property. Thus, we guess the amount of
+    Lexicons we are going to process based on Lexicon and
+    LexiconExtension tag.
+    """
+    iterations = 0
+    with open(source, "rb") as fp:
+        for line in fp:
+            if LEXICON_TAG in line or LEXICON_EXT_TAG in line:
+                iterations += 1
+    return iterations
```
[the remark this answers is not in the document]
Comment (dpalmasan) [reply wn/lmf.py]: Addressed, you were right regarding this one, my bad.

File: wn/lmf.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dpalmasan) [reply wn/lmf.py]: My bad, I also addressed this comment, now the progress update occur on reading a lexical entity and not the complete lexicon as I previously did.

File: wn/_add.py [hunk 3]
```diff
@@ -73,6 +74,7 @@
 def add(
     source: AnyPath,
     progress_handler: Optional[Type[ProgressHandler]] = ProgressBar,
+    lmf_load_progress: Optional[Type[ProgressHandler]] = ProgressBar,
```
Comment (goodmami) [inline wn/_add.py:77]: The pattern I used for `wn.download()` is that the value of `progress_handler` is passed to `wn.add()`. I think the same could be done here. That means the *class* and not the instance is passed along.

File: wn/_add.py [hunk 4]
```diff
@@ -90,14 +92,18 @@ def add(
         progress_handler = ProgressHandler
     progress = progress_handler(message='Database')
 
+    if lmf_load_progress is None:
+        lmf_load_progress = ProgressHandler
+    progress_lmf_load = lmf_load_progress(message='XML Read')
```
Comment (goodmami) [inline wn/_add.py:97]: Why is the LMF progress bar being instantiated here?

File: wn/_add.py [hunk 5]
```diff
@@ -138,10 +145,13 @@ def _add_lmf(
         _update_lookup_tables(all_infos, cur)
 
         progress.flash(f'Reading {source!s}')
-        for lexicon, info in zip(lmf.load(source), all_infos):
+        gen_progress = GeneratorProgressInfo(
+            total_items=sum(info["counts"]["LexicalEntry"] for info in all_infos),
```
Comment (goodmami) [inline wn/_add.py:149]: Why only lexical entries? The lexicon contains much more than that.

File: wn/lmf.py [hunk 6]
```diff
@@ -90,12 +99,31 @@ def __iter__(self):
...
             raise StopIteration
         self._next = next(self.iterator, (None, None))
         return _next
 
+    def _update_progress(self):
+        if self._progress_info:
+            self._progress += 1
+            # We don't want to refresh each cycle
+            refresh_cycle = 20
+            if self._progress % refresh_cycle == 0:
```
Comment (goodmami) [inline wn/lmf.py:113]: I understand the performance reasons why we don't refresh at each iteration, but I think it'd be cleaner to move this logic to the `wn.util.ProgressHandler` and `ProgressBar` classes. E.g.,

```diff
class ProgressHandler:
    ...
    def __init__(
        self,
        *,
        message: str = '',
        count: int = 0,
        total: int = 0,
+       refresh_interval: int = 1,
        unit: str = '',
        status: str = '',
        file: TextIO = sys.stderr,
    ):
        self.file = file
        self.kwargs = {
            'count': count,
            'total': total,
+           'refresh_interval': refresh_interval,
            'message': message,
            'unit': unit,
            'status': status,
        }
+        self._refresh_quota: int = refresh_interval
...
class ProgressBar(ProgressHandler):
    ...
    def update(self, n: int = 1) -> None:
        """Increment the count by *n* and print the reformatted bar."""
        self.kwargs['count'] += n  # type: ignore
+       self._refresh_quota -= n
+       if self._refresh_quota <= 0:
+           self._refresh_quota = self.kwargs['refresh_interval']
+           s = self.format()
-       s = self.format()
-       ...
```

File: wn/lmf.py [hunk 7]
```diff
@@ -90,12 +99,31 @@ def __iter__(self):
...
+                    self._refresh_count += refresh_cycle
+                else:
+                    last_refresh = self._refresh_count
+                    last_refresh = self._progress_info.total_items - last_refresh
+                    self._progress_info.progress_handler.update(last_refresh)
+
     def starts(self, *tags: str) -> bool:
         event, elem = self._next
+        if "LexicalEntry" in tags:
+            self._update_progress()
```
Comment (goodmami) [inline wn/lmf.py:125]: Again, I think this should be updated for more than just lexical entries.

File: wn/lmf.py [hunk 8]
```diff
@@ -482,7 +510,7 @@ def __init__(
         self.requires = requires or []
 
 
-LexicalResource = List[Lexicon]
+LexicalResource = Iterator[Lexicon]
```
Comment (goodmami) [inline wn/lmf.py:513]: Why is this change happening (here and throughout)? It seems unrelated to the addition of a progress bar.

File: wn/util.py [hunk 9]
```diff
@@ -153,3 +153,9 @@ def flash(self, message: str) -> None:
     def close(self) -> None:
         """Print a newline so the last printed bar remains on screen."""
         print(file=self.file)
+
+
+class GeneratorProgressInfo(NamedTuple):
+    """Progress Bar info for generator."""
+    total_items: int
+    progress_handler: ProgressHandler
```
Comment (goodmami) [inline wn/util.py:161]: I'm not certain I see the need for this new class. What is it's purpose?

**Review by goodmami — CHANGES_REQUESTED** [review]
Thanks, this is much better. It still needs some work, though:
* The change to having `wn.lmf.load()` yield lexicons instead of returning a list of them is superfluous. I don't think it's a bad idea in general, but it should go in its own issue and PR.
* Only lexical entries are counted for some reason.
* The logic for when to refresh the progress bar is not contained within the progress handler classes. I try to keep the messaging code (logging, progress, etc.) to a minimum in the "business" code. This also precludes someone using a custom progress bar from choosing their own refresh settings.

Also, to avoid wasting effort, feel free to ask questions before going off to code the next iteration, if you have any.

File: wn/util.py (hunk 9 above)
Comment (dpalmasan) [reply wn/util.py:161]: The purpose is just holding the progress_handler and the total count to pass them downstream to the load function. Another way could just be changing again the signature of `lmf.load`, and pass the two arguments. But I don't think that will look clean.

Comment (dpalmasan) [reply wn/util.py:161]: Actually my bad, I could have set the `ProgressHandler` before even calling `load`

Comment (goodmami) [reply wn/util.py:161]: Ok yes, I see the issue: `lmf.load()` doesn't call `lmf.scan_lexicons()` itself so it doesn't have those counts. I think, for now, what you describe is sufficient: instantiating a ProgressHandler and passing the object (instead of the class) to `load()`. It breaks the current pattern, but the other options require too much work for this PR.

Comment (dpalmasan) [reply wn/util.py]: Yes, but there is a solution, I think it might be cleaner than creating the class I was creating. I just set the progress before calling `lmf.load`. I think I have addressed this to maintain the pattern

File: wn/lmf.py (hunk 7 above)
Comment (dpalmasan) [reply wn/lmf.py]: addressed

File: wn/lmf.py (hunk 6 above)
Comment (dpalmasan) [reply wn/lmf.py]: Thanks, this one is pretty good actually.

File: wn/_add.py [hunk 10]
```diff
@@ -138,10 +140,23 @@ def _add_lmf(
...
+        if lmf_progress_handler is None:
+            lmf_progress_handler = ProgressHandler
+        progress_lmf = lmf_progress_handler(message='XML read')
+
+        total_items = sum(_sum_counts(info) for info in all_infos)
+        # Since we are going to call update on __next__ of the XML
+        # Event Iterator, we will see the tags twice, because we
+        # see the opening and closing tags. This is why, the total
+        # For the lmf progress handled is the double of the total
+        # sum.
```
Comment (goodmami) [inline wn/_add.py:153]: Interesting. Alternatively, if someone cares about the actual count, we could just update on the "end" events.

File: wn/_add.py [hunk 11]
```diff
@@ -73,6 +73,7 @@
 def add(
     source: AnyPath,
     progress_handler: Optional[Type[ProgressHandler]] = ProgressBar,
+    lmf_progress_handler: Optional[Type[ProgressHandler]] = ProgressBar,
```
Comment (goodmami) [inline wn/_add.py:76]: I still don't like the idea of having a separate progress handler class. I think it is unlikely that a user will want a separate class (as opposed to separate instances) for each individual process. Let's just reuse `progress_handler` to instantiate the `progress_lmf` below.

**Review by goodmami — CHANGES_REQUESTED** [review]
Even better! I have a few minor quibbles, but otherwise I think this is close to being mergeable.

File: wn/_add.py (hunk 11 above)
Comment (dpalmasan) [reply wn/_add.py:76]: You are totally right.

File: wn/util.py [hunk 12]
```diff
@@ -102,11 +104,14 @@ class ProgressBar(ProgressHandler):
 
     def update(self, n: int = 1) -> None:
         """Increment the count by *n* and print the reformatted bar."""
+        self._refresh_quota -= n
+        if self._refresh_quota <= 0:
+            self._refresh_quota = self.kwargs['refresh_interval']  # type: ignore
+            s = self.format()
+            if self.file:
+                print('\r\033[K', end='', file=self.file)
+                print(s, end='', file=self.file)
```
[the remark this answers is not in the document]
Comment (dpalmasan) [reply wn/util.py:115]: Addressed

File: wn/_add.py (hunk 10 above)
Comment (dpalmasan) [reply wn/_add.py]: Okay I swear I tried that but got the wrong counts. But maybe it was in the previous version of the code in which I wasn't considering the full count. I moved and addressed the comment, now it shows the actual count, not the hacky 2X

File: wn/_add.py [hunk 13]
```diff
@@ -138,10 +138,16 @@ def _add_lmf(
         _update_lookup_tables(all_infos, cur)
 
         progress.flash(f'Reading {source!s}')
-        for lexicon, info in zip(lmf.load(source), all_infos):
+
+        total_items = sum(_sum_counts(info) for info in all_infos)
+        progress.set(count=0,
+                     total=total_items,
+                     refresh_interval=50,
+                     status='Reading XML')
```
Comment (goodmami) [inline wn/_add.py:146]: Not exactly what I meant. I meant to use the same `ProgressHandler` class to create the `progress_lmf` instance. But let's just leave this for now as it's not so bad to reuse the same progress bar.

File: wn/util.py [hunk 14]
```diff
@@ -103,10 +105,14 @@ class ProgressBar(ProgressHandler):
     def update(self, n: int = 1) -> None:
         """Increment the count by *n* and print the reformatted bar."""
         self.kwargs['count'] += n  # type: ignore
-        s = self.format()
-        if self.file:
-            print('\r\033[K', end='', file=self.file)
-            print(s, end='', file=self.file)
+        self._refresh_quota -= n
+        draw_bar = self._refresh_quota <= 0
+        if draw_bar:
```
Comment (goodmami) [inline wn/util.py:110]: Why not just the following?

```python
if self._refresh_quota <= 0:
```

I don't think we need to declare a new variable to hold a one-off comparison value.

**Review by goodmami — CHANGES_REQUESTED** [review]
Looks good. Let's fix the `draw_bar` thing (see below) and maybe the 2x event issue, then I can approve it.

File: wn/util.py (hunk 14 above)
Comment (dpalmasan) [reply wn/util.py]: Sorry about that, it was kind of a leftover of something silly I have done previously. Addressed.

File: wn/_add.py (hunk 13 above)
Comment (dpalmasan) [reply wn/_add.py]: addressed. I instantiated using the same class, and created another progress handler for the lmf load.

**Review by goodmami — APPROVED** [review]
Ok that looks pretty good.

I'll approve and merge into the next release branch. I think it needs a bit of tweaking for performance reasons, though. With a `refresh_interval` of 50, it adds increases the time to add the `ewn:2020` lexicon by 6s (from 39s to 45s). If I change the interval to 1000, it increases the time by 3s. The EWN lexicon has about 1.25 million elements, and since the default bar width is 30, and there are 3 steps at each column (`-`, `=`, and `#`), we could make the interval a bit higher than 10000 and still see a smooth progression (of course, smaller lexicons have many fewer elements). When I try 10000 as the interval, I didn't seem much slowdown at all.

But don't worry about the performance issue.. I'll adjust that in a separate commit.

**Comment by goodmami** [comment]
Thanks for your help!
